File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/ui/views/MTGLoader.java
package com.dbottillo.mtgsearchfree.ui.views;


import android.animation.ArgbEvaluator;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.os.Handler;
import android.os.Message;
import android.support.v4.view.animation.FastOutLinearInInterpolator;
import android.util.AttributeSet;
import android.view.View;

import com.dbottillo.mtgsearchfree.R;

import java.lang.ref.WeakReference;

public class MTGLoader extends View {

    private int size;
    private float maxRadius;
    private float currentStep = 0.0f;
    private int currentPaintColor = 0;
    private boolean growing = true;
    private final static float STEP = 0.05f;
    private final static int DELAY_MILLIS = 40;
    private ArgbEvaluator argbEvaluator;
    private FastOutLinearInInterpolator interpolator;
    private Paint paint = new Paint();

    public MTGLoader(Context context) {
        super(context);
        init();
    }

    public MTGLoader(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public MTGLoader(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        argbEvaluator = new ArgbEvaluator();
        interpolator = new FastOutLinearInInterpolator();
        setPaintColor();
        MTGLoaderHandler loadingHandler = new MTGLoaderHandler(this);
        loadingHandler.sendEmptyMessage(0);
    }

    private void setPaintColor() {
        paint.setColor(getCurrentColor());
    }

    private int getCurrentColor() {
        return getColor(currentPaintColor);
    }

    private int getNextColor() {
        int next = currentPaintColor + 1;
        if (next > 4) {
            next = 0;
        }
        return getColor(next);
    }

    private int getColor(int value) {
        Resources resources = getResources();
        if (value == 0) {
            return resources.getColor(R.color.mtg_white);
        }
        if (value == 1) {
            return resources.getColor(R.color.mtg_blue);
        }
        if (value == 2) {
            return resources.getColor(R.color.mtg_black);
        }
        if (value == 3) {
            return resources.getColor(R.color.mtg_red);
        }
        if (value == 4) {
            return resources.getColor(R.color.mtg_green);
        }
        return -1;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        float value = interpolator.getInterpolation(currentStep);

        if (!growing) {
            int color = (Integer) argbEvaluator.evaluate(value, getCurrentColor(), getNextColor());
            paint.setColor(color);
            value = 1.0f - value;
        }

        int radius = (int) (maxRadius * value);
        canvas.drawCircle(size / 2.f,  size / 2.f, radius, paint);

        currentStep += STEP;
        if (currentStep >= 1.0f) {
            if (growing) {
                growing = false;
            } else {
                currentPaintColor++;
                if (currentPaintColor > 4) {
                    currentPaintColor = 0;
                }
                growing = true;
            }
            currentStep = 0;
        }

    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, widthMeasureSpec); // always want a square canvas

        size = getMeasuredWidth();
        maxRadius = size / 6.f;
    }

    private static class MTGLoaderHandler extends Handler {

        WeakReference<MTGLoader> loader;

        MTGLoaderHandler(MTGLoader loader) {
            this.loader = new WeakReference<>(loader);
        }

        @Override
        public synchronized void handleMessage(Message msg) {
            MTGLoader mtgLoader = loader.get();
            if (mtgLoader != null) {
                mtgLoader.invalidate();
                sendEmptyMessageDelayed(0, DELAY_MILLIS);
            }
        }
    }

}

File: MTGSearch/src/main/kotlin/com/dbottillo/mtgsearchfree/ui/views/MTGLoader.kt
package com.dbottillo.mtgsearchfree.ui.views

import android.animation.ArgbEvaluator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.os.Handler
import android.os.Message
import android.support.v4.content.ContextCompat
import android.support.v4.view.animation.FastOutLinearInInterpolator
import android.util.AttributeSet
import android.view.View
import com.dbottillo.mtgsearchfree.R
import java.lang.ref.WeakReference

class MTGLoader : View {

    private var size: Int = 0
    private var maxRadius: Float = 0.toFloat()
    private var currentStep = 0.0f
    private var currentPaintColor = 0
    private var growing = true
    private var argbEvaluator = ArgbEvaluator()
    private var interpolator = FastOutLinearInInterpolator()
    private val paint = Paint()

    private val currentColor: Int
        get() = getColor(currentPaintColor)

    private val nextColor: Int
        get() {
            var next = currentPaintColor + 1
            if (next > 4) {
                next = 0
            }
            return getColor(next)
        }

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    private fun init() {
        paint.isAntiAlias = true
        setPaintColor()
        val loadingHandler = MTGLoaderHandler(this)
        loadingHandler.sendEmptyMessage(0)
    }

    private fun setPaintColor() {
        paint.color = currentColor
    }

    private fun getColor(value: Int): Int {
        return ContextCompat.getColor(context, when (value) {
            0 -> R.color.mtg_white
            1 -> R.color.mtg_blue
            2 -> R.color.mtg_black
            3 -> R.color.mtg_red
            4 -> R.color.mtg_green
            else -> -1
        })
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        var value = interpolator.getInterpolation(currentStep)

        if (!growing) {
            val color = argbEvaluator.evaluate(value, currentColor, nextColor) as Int
            paint.color = color
            value = 1.0f - value
        }

        val radius = (maxRadius * value).toInt()
        canvas.drawCircle(size / 2f, size / 2f, radius.toFloat(), paint)

        currentStep += STEP
        if (currentStep >= 1.0f) {
            if (growing) {
                growing = false
            } else {
                currentPaintColor++
                if (currentPaintColor > 4) {
                    currentPaintColor = 0
                }
                growing = true
            }
            currentStep = 0f
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, widthMeasureSpec) // always want a square canvas
        size = measuredWidth
        maxRadius = size / 6f
    }

    private class MTGLoaderHandler internal constructor(loader: MTGLoader) : Handler() {

        private var loader: WeakReference<MTGLoader> = WeakReference(loader)

        @Synchronized override fun handleMessage(msg: Message) {
            val mtgLoader = loader.get()
            if (mtgLoader != null) {
                mtgLoader.invalidate()
                sendEmptyMessageDelayed(0, DELAY_MILLIS.toLong())
            }
        }
    }

    companion object {
        private const val STEP = 0.05f
        private const val DELAY_MILLIS = 40
    }
}